docs/usecases/streaming.md

```diff
@@ -3,4 +3,105 @@ id: usecases_streaming
+is not thread-safe so it should not be evaluated from multiple fibers. The evaluation also cannot be safely restarted after 
+interruption in all cases. When the `Pull` fails with a `None` it is not safe to evaluate it again. All of those seem
+very restrictive but keeping in mind that a stream will talk to external world possibly multiple times, which can be
+stateful, those assumptions are the only safe ones we can make. To give a concrete example, if our stream job is to
+interact with an external API that has time throttling of request it can accept from a single user then even if our 
+stream implementation does proper buffering and back-off in presence of two concurrent not coordinated clients both of
+them will probably end up with failures. Other example for getting `None` as error value. If our stream is responsible 
+for reading a file, `None` represents end of the file. Trying to pull more values after the file was wholly read makes
+no sense.
+
+The last type widely used in `zio-stream` is the `ZSink` type. Sink should be understood as a place that runs the stream
+and where the evaluated items are being passed down to. Sinks also have control over the evaluation of streams, possibly
+ending them earlier. `Sink[-R, +E, +A0, -A, +B]` contract states that sinks consume multiple values of type `A` coming
+from the `ZStream` type, producing either a single value of type B and reminders of type `A0` or an error of type `E`. 
+If we recall the `ZStream` and `Pull` definitions we can fast recognize that there is not space of "state" in both of
+them. This responsibility is being passed on to `ZSink` type. Example where we need such state would be a scenario where
+we want to count the number of elements that flow though the stream. Here we're moving out of the simple scenario of 
+transforming the incoming data into, keeping derived data based on the history of execution. Thus to support the need
+for such operators `ZSink` has a standardized interface to make this possible. First of all any sink may have some kind of
+state. In the general case `State` is a path-dependent type defined in `ZSink` that is hidden from the client of 
+the stream and might be anything that is needed so it is fully abstract. Next there are number of methods exposed on 
+`ZSink`. Let's enumerate them with simplified types:
+- `initial` - returns initial `state` of the stream which might be anything
+- `step` - is a function `(state, a) => state`. Step takes the current state, incoming item and calculates the next state.
+- `extract`: `state => (B, Chunk[A0])`, calculates the final value that is the result of running the stream and list of 
+  values that were not consumed.
+- `cont`: `state => boolean`, controls should the stream be evaluated further.
+
+Let's go through the `run` definition on `ZStream` which allows for connecting any `ZSink` to have a better 
```

Comment:
I'm a bit wary of going through the definition of `run` here because it will undoubtedly change. Maybe it'd be better to conceptually describe how `run` consumes a stream using a sink?